crates/ee-acct-runtime/src/process_update.rs

```diff
@@ -0,0 +1,186 @@
+//! High-level procedures for processing updates.
+//!
+//! There are two main functions here.  The first is
+//! [`verify_and_apply_update_operation`], which is what we use in the proof to
+//! carefully check and update the account state.
+//!
+//! The second is [`apply_update_operation_unconditionall`], which is used
+//! outside the proof, after verifying the proof, to update our view of the
+//! state, presumably with information extracted from DA.
+
+use strata_acct_types::AcctId;
+use strata_ee_acct_types::{
+    CommitCoinput, DecodedEeMessage, EeAccountState, EnvError, EnvResult, ExecutionEnvironment,
+};
+use strata_snark_acct_types::{MessageEntry, UpdateOperation};
+
+use crate::{
+    exec_processing::{apply_commit, verify_commit},
+    verification_state::UpdateVerificationState,
+};
+
+/// Verify if an update operation is valid.  Accepts coinputs corresponding to
+/// each message to privately attest validity before applying effects.
+pub fn verify_and_apply_update_operation<'i, E: ExecutionEnvironment>(
+    state: &mut EeAccountState,
+    operation: &UpdateOperation,
+    coinputs: impl IntoIterator<Item = &'i [u8]>,
+    ee: &E,
+) -> EnvResult<()> {
+    let mut coinp_iter = coinputs.into_iter();
+
+    // 1. Process each message in order.
+    let mut vstate = UpdateVerificationState::new_from_state(state);
+    for (inp, coinp) in operation.processed_messages().iter().zip(&mut coinp_iter) {
+        let Some((meta, eem)) = parse_input(&inp) else {
+            // Other type or invalid message, skip.
+            continue;
+        };
+
+        // Verify the message.  This relies on the private coinput.
+        verify_message(&mut vstate, state, &eem, &meta, coinp, operation, ee)?;
+
+        // Then apply the message.  This doesn't rely on the private coinput.
+        apply_message(state, &eem, &meta, operation)?;
+    }
```

Comment:
This is not an atomic operation.
What happens if we process multiple `(inp, coinp)`s and one or more of these are invalid. Then the `EeAccountState` would be changed in the loop even if the `verify_accumulated_state` below fails.

Is this expected? If not, we should accumulate all the changes in the `EeAccountState`, and then apply them after the `verify_accumulated_state` below to make this `verify_and_apply_update_operation` atomic and without side effects if it fails.

Reply:
This state is small and in memory, so we're not modifying an "authoritative" instance.  We'd discard the modified state if the function returns an err.  From the context of the proof, the err paths are never taken.

Maybe `verify_message` is a misleading name.  It's not verifying the *message* per se, it's challenging the prover to potentially "do something with" the message (via the coinput).  If they can't then they can't construct the proof.

Reply:
Solving, thanks for the explanation